### PR TITLE
Feature/local development

### PR DESCRIPTION
**Pull request by Rampage1xx**
Hello everyone, 
With this PR I aim to introduce a local development playground that is easy to spin-up and to work with.

Design principles:

- enable local development with least amount of code needed
- do things in an unopinionated way. This PR aims to not alter the structure of the project in any way. Just do one thing and do it right.
- Any alternation needed for the local development to work are code insertion that work only if the local development flag is active. This is to avoid subtle breaking changes in production. This PR does not aim to refactor code. The only exception to this rule is the centralized  httpClient due to being very safe to implement with the added benefit to avoid a lot of dirty code inside the frontend codebase

Cliff notes:

- split frontend and backend containers in local env
- local env has file watchers that rebuild the frontend or backend package if there are changes
- no more multiple axios instances in the frontend. There is only one axios instane and is exported as httpClient in the frontend. This is needed to propagate the local backend baseUrl in a clean way, it will also be needed to implement basicAuth in local development in the future
- fixed inconsistencies in the makefile
- replaced npm with yarn for it's deterministic/retro-compatible approach
- fixes pipelines
- disables basicAuth and static react bundle in local environment.
- Tested on OSX. Still untested on UNIX. 

You are **strongly encouraged to test** before LGTM

**DO NOT SQUASH COMMITS WHEN MERGING**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,25 +22,35 @@ ifndef GIT_UNCLEAN
 	$(if $(shell git status -s), $(error Git is tainted:: $(shell git status -s) ***),)
 endif
 
+# Local Development
+
+.PHONY: development-start
+development-start:
```
Comment (gabrielelana) [inline Makefile:28]: Excuse my OCD, just nitpicking: `start` and `stop` or `up` and `down`

File: development/dev-start.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+kind create cluster --config=./development/kind-config.yml --kubeconfig=./.kubeconfig
+# we make the kubeconfig for the local backend
+cp .kubeconfig .kubeconfig-backend
+# containers reach the host on the 'host.docker.internal' dns instead of 127.0.0.1
+# this line is needed for cross compatibility between osx and unix
+#sed -i.bak 's/127.0.0.1/host.docker.internal/g' .kubeconfig-backend && rm -f .kubeconfig-backend.bak
+sed -i.bak "s/server: $PARTITION_COLUMN.*/server: https:\/\/kind-control-plane:6443/" .kubeconfig-backend && rm -f .kubeconfig-backend.bak
+
+# now that there is the .kubeconfig we can evaluate the .envrc
+source .envrc
```
Comment (gabrielelana) [inline development/dev-start.sh:14]: Since `.envrc` is going to be automatically sourced by everybody that have `direnv` installed on their machine and this environment file is supposed to be sourced only after the cluster has started, can we rename the file?

File: docs/how-to-contribute.md [hunk 3]
```diff
@@ -49,10 +50,10 @@ To create a local kind cluster, run the `make kind-cluster` command.
 
 #### TL;DR
 ```
-kind create cluster
```
Comment (gabrielelana) [inline docs/how-to-contribute.md:52]: Shouldn't this be replaced with `make development-start`?

File: docs/local-development.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Local Development
+
+## Requirements
+
+- docker v20.10. Required for host.docker.internal on UNIX systems
```
Comment (gabrielelana) [inline docs/local-development.md:5]: `docker-compose` is missing in the list

Comment (gabrielelana) [reply docs/local-development.md:5]: What's the difference between this list and the list in `how-to-contribute.yml`?

File: development-compose.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      context: .
+    volumes:
+      - ./:/app
+      - ./.kubeconfig-backend:/root/.kube/config
+    networks:
+      - kind
+
+networks:
+  kind:
+    external: true
```
Comment (angelbarrera92) [inline development-compose.yml:37]: Nice hack!

File: docs/how-to-contribute.md [hunk 6]
```diff
@@ -23,13 +23,14 @@ Note that in a future version of the software, the current naming convention wil
...
 - nodejs (developed on v13.1.0)
-- [kind](https://github.com/kubernetes-sigs/kind)
+- [kind](https://github.com/kubernetes-sigs/kind) (v0.9.0)
 
 Required packages
 
 - make 4.1
 - golang 1.14.2
 - npm 6.14.4
+- yarn 1.22.10
```
Comment (gabrielelana) [inline docs/how-to-contribute.md:33]: Also here I guess `bats` is missing

**Review by gabrielelana — CHANGES_REQUESTED** [review]
And in the end... it works on Linux with Docker 20.10 👍 

Only a couple of end2end tests failed but I think this is a good starting point to revive the project

File: docs/how-to-contribute.md (hunk 3 above)
Comment (Rampage1xx) [reply docs/how-to-contribute.md:52]: how-to-contribute.md gives instruction on how to build a permission-manager image and deploy it to kubernetes cluster.
 It is different from the local development one

File: development/dev-start.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+set -e
+# we create the kind cluster with some specific configurations, ex: we allow 'host.docker.internal' on the kubeconf
+# we also save the .kubeconfig on the root folder
+kind create cluster --config=./development/kind-config.yml --kubeconfig=./.kubeconfig
+# we make the kubeconfig for the local backend
+cp .kubeconfig .kubeconfig-backend
+# containers reach the host on the 'host.docker.internal' dns instead of 127.0.0.1
+# this line is needed for cross compatibility between osx and unix
+#sed -i.bak 's/127.0.0.1/host.docker.internal/g' .kubeconfig-backend && rm -f .kubeconfig-backend.bak
```
[the remark this answers is not in the document]
Comment (Rampage1xx) [reply development/dev-start.sh:10]: Another line was needed to make it work on UNIX, I will delete it

File: docs/local-development.md (hunk 4 above)
Comment (Rampage1xx) [reply docs/local-development.md:5]: how-to-contribute.md gives instruction on how to build a permission-manager image and deploy it to kubernetes cluster.

It is different from the local development one

It should be given better names to the files IMO

**Comment by Rampage1xx** [comment]
can you provide more info on failed e2e tests?

**Comment by gabrielelana** [comment]
> can you provide more info on failed e2e tests?

There you go

```console
make test-e2e
1..2
1: [SETUP] Deploy Permission manager Requirements
ok 1 [SETUP] Deploy Permission manager Requirements
2: [SETUP] Deploy Permission manager
ok 2 [SETUP] Deploy Permission manager
1..7
1: [USERS] Create test-user
ok 1 [USERS] Create test-user
2: [USERS] Create role-binding to test-user
ok 2 [USERS] Create role-binding to test-user
3: [USERS] Create cluster-role-binding to test-user
ok 3 [USERS] Create cluster-role-binding to test-user
4: [USERS] Create kubeconfig to test-user
ok 4 [USERS] Create kubeconfig to test-user
5: [USERS] Modify kubeconfig (insecure-skip-tls-verify
ok 5 [USERS] Modify kubeconfig (insecure-skip-tls-verify
6: [USERS] Test generated kubeconfig - test-user can not deploy pods in kube-system
can-i create pods -n kube-system: Warning: the server doesn't have a resource type 'pods'
The connection to the server localhost:8080 was refused - did you specify the right host or port?
not ok 6 [USERS] Test generated kubeconfig - test-user can not deploy pods in kube-system
# (in test file tests/create-user.sh, line 63)
#   `[ "$status" -eq 0 ]' failed
7: [USERS] Test generated kubeconfig - test-user can deploy pods in permission-manager
can-i create pods -n permission-manager: Warning: the server doesn't have a resource type 'pods'
The connection to the server localhost:8080 was refused - did you specify the right host or port?
not ok 7 [USERS] Test generated kubeconfig - test-user can deploy pods in permission-manager
# (in test file tests/create-user.sh, line 74)
#   `[ "$status" -eq 0 ]' failed
make: *** [Makefile:69: test-e2e] Error 1
```

**Comment by Rampage1xx** [comment]
Cannot reproduce :( everything working on OSX. I will try to dig deeper

1..2
1: [SETUP] Deploy Permission manager Requirements
ok 1 [SETUP] Deploy Permission manager Requirements
2: [SETUP] Deploy Permission manager
ok 2 [SETUP] Deploy Permission manager
1..7
1: [USERS] Create test-user
{"name":"test-user"}
ok 1 [USERS] Create test-user
2: [USERS] Create role-binding to test-user
{"ok":true}
ok 2 [USERS] Create role-binding to test-user
3: [USERS] Create cluster-role-binding to test-user
{"ok":true}
ok 3 [USERS] Create cluster-role-binding to test-user
4: [USERS] Create kubeconfig to test-user
{"ok":true,"kubeconfig":"---\napiVersion: v1\nkind: Config\ncurrent-context: test-user@kind-kind\nclusters:\n  - cluster:\n     }ok 4 [USERS] Create kubeconfig to test-user
5: [USERS] Modify kubeconfig (insecure-skip-tls-verify
ok 5 [USERS] Modify kubeconfig (insecure-skip-tls-verify
6: [USERS] Test generated kubeconfig - test-user can not deploy pods in kube-system
can-i create pods -n kube-system: no
ok 6 [USERS] Test generated kubeconfig - test-user can not deploy pods in kube-system
7: [USERS] Test generated kubeconfig - test-user can deploy pods in permission-manager
can-i create pods -n permission-manager: yes
ok 7 [USERS] Test generated kubeconfig - test-user can deploy pods in permission-manager
Forwarding from 127.0.0.1:4000 -> 4000
Forwarding from [::1]:4000 -> 4000

**Comment by Rampage1xx** [comment]
@gabrielelana  issues should(hopefully) be fixed now

File: Makefile [hunk 8]
```diff
@@ -24,9 +24,9 @@ endif
 
 # Local Development
 
-.PHONY: development-start
-development-start:
-	@./development/dev-start.sh
+.PHONY: development-up
+development-up:
+	@./development/dev-up.sh
```
Comment (gabrielelana) [inline Makefile:29]: Nice 👍

File: development/dev-up.sh [hunk 9]
```diff
@@ -7,11 +7,10 @@ kind create cluster --config=./development/kind-config.yml --kubeconfig=./.kubec
 cp .kubeconfig .kubeconfig-backend
 # containers reach the host on the 'host.docker.internal' dns instead of 127.0.0.1
 # this line is needed for cross compatibility between osx and unix
-#sed -i.bak 's/127.0.0.1/host.docker.internal/g' .kubeconfig-backend && rm -f .kubeconfig-backend.bak
 sed -i.bak "s/server: $PARTITION_COLUMN.*/server: https:\/\/kind-control-plane:6443/" .kubeconfig-backend && rm -f .kubeconfig-backend.bak
 
-# now that there is the .kubeconfig we can evaluate the .envrc
-source .envrc
+# now that there is the .kubeconfig we can evaluate the .env-cluster
```
Comment (gabrielelana) [inline development/dev-up.sh:12]: Nice 👍

**Review by gabrielelana — APPROVED** [review]
❤️ ❤️ ❤️

**Comment by gabrielelana** [comment]
> Cannot reproduce :( everything working on OSX. I will try to dig deeper

@Rampage1xx ping me when you have time, we can look at it together
